vllm/entrypoints/openai/logits_processors.py

```diff
@@ -0,0 +1,75 @@
+from functools import lru_cache
+from typing import Dict, FrozenSet, Iterable, List, Optional, Union
+
+import torch
+from transformers import PreTrainedTokenizer
+
+from vllm.sampling_params import LogitsProcessor
+
+
+class AllowedTokenIdsLogitsProcessor:
+    """Logits processor for constraining generatedtokens to a
+    specific set of token ids."""
+
+    def __init__(self, allowed_ids: Iterable[int]):
+        self.allowed_ids: Optional[List[int]] = list(allowed_ids)
+        self.mask: Optional[torch.Tensor] = None
+
+    def __call__(self, token_ids: List[int],
+                 logits: torch.Tensor) -> torch.Tensor:
+        if self.mask is None:
+            self.mask = torch.full((logits.shape[-1], ),
+                                   True,
+                                   dtype=torch.bool,
+                                   device=logits.device)
```

Comment:
Is it guaranteed that logits will always be from the same device even for multi-gpu inference?

Reply:
@DarkLight1337 yes I think this is safe to assume. There's never more than one GPU device used per process AFAIK.